rust/lance/src/io/commit/conflict_resolver.rs

```diff
@@ -827,13 +1002,152 @@ impl<'a> TransactionRebase<'a> {
+                | Operation::UpdateMemWalState { .. } => Ok(()),
             }
         } else {
             Err(wrong_operation_err(&self.transaction.operation))
         }
     }
 
+    fn check_update_mem_wal_state_txn(
+        &mut self,
+        other_transaction: &Transaction,
+        other_version: u64,
+    ) -> Result<()> {
+        if let Operation::UpdateMemWalState {
+            added,
+            updated,
+            removed,
+            ..
+        } = &self.transaction.operation
+        {
+            match &other_transaction.operation {
+                Operation::UpdateMemWalState {
+                    added: committed_added,
+                    updated: committed_updated,
+                    removed: committed_removed,
+                } => {
+                    // 1. if the current or last committed job is trimming flushed MemWALs,
+                    // it is compatible with any other UpdateMemWalState commits
+                    if (committed_added.is_empty()
+                        && committed_updated.is_empty()
+                        && !committed_removed.is_empty())
+                        || (added.is_empty() && updated.is_empty() && !removed.is_empty())
+                    {
+                        return Ok(());
+                    }
+
+                    // 2. MemWALs of different regions can be changed at the same time
+                    self.check_update_mem_wal_state_not_modify_same_mem_wal(
+                        committed_added,
+                        added,
+                        other_transaction,
+                        other_version,
+                    )?;
+                    self.check_update_mem_wal_state_not_modify_same_mem_wal(
+                        committed_added,
+                        updated,
+                        other_transaction,
+                        other_version,
+                    )?;
+                    self.check_update_mem_wal_state_not_modify_same_mem_wal(
+                        committed_updated,
+                        added,
+                        other_transaction,
+                        other_version,
+                    )?;
+                    self.check_update_mem_wal_state_not_modify_same_mem_wal(
+                        committed_updated,
+                        updated,
+                        other_transaction,
+                        other_version,
+                    )?;
+                    Ok(())
+                }
+                Operation::Update {
+                    mem_wal_to_flush, ..
+                } => {
+                    if mem_wal_to_flush.is_some() {
+                        Ok(())
+                    } else {
+                        Err(self.incompatible_conflict_err(
+                            other_transaction,
+                            other_version,
+                            location!(),
+                        ))
+                    }
```

Comment:
Is this if/else backward?

Or is the thought that an update that is flushing a mem-wal is making no other changes?

Reply:
Yes "an update that is flushing a mem-wal is making no other changes".

Reply:
The check could be more detailed, there is an assumption that once a MemWAL is sealed, there is no other operation that could change the state back to open, and at that point it can always be flushed.